### ISSUE TITLE
'is same as(0.0)' evaluates to false with 0.0 value

### ISSUE THREAD

**Issue by josefvn** [twigphp/Twig#2788]
Refer to https://twigfiddle.com/agvgz9/2

If we have the following context:

```
v1: 0.0
v2: 0.0
```
The following Twig snippet:

```
{{ 0.0 is same as (0.0) ? 'Yes' : 'No' }}
{{ v1 is same as (0.0) ? 'Yes' : 'No' }} 
{{ v1 is same as (v2) ? 'Yes' : 'No' }}
```

results in the following, respectively:

```
Yes
No
Yes
```

One would expect line 2 to evaluate to true as well, as you would find in PHP:

<img width="640" alt="screen shot 2018-12-20 at 10 40 25" src="https://user-images.githubusercontent.com/612695/50282064-e1e11f00-0451-11e9-9c7b-14089701d472.png">

**Comment by stof** [comment]
The issue here is that `0.0` gets turned into `0` at some point, and so does not stay a float in the compiled template.

**Comment by DarkBee** [comment]
Yes we figured that out already as seen in the [comments](https://stackoverflow.com/questions/53866084/twig-is-same-as0-0-for-the-value-of-0-0-is-false). We were wondering why twig actualy compiles the `(float)0.0` into `(int)0`

**Comment by stof** [comment]
Well, the comments you are linking about is for a StackOverflow question which was not referenced in the issue at all. So I had no way to know about them.

**Comment by fabpot** [comment]
Just tested the following with both 1.x and 2.x:

```twig
{% set v1 = 0.0 %}
{% set v2 = 0.0 %}

{{ 0.0 is same as (0.0) ? 'Yes' : 'No' }}
{{ v1 is same as (0.0) ? 'Yes' : 'No' }}
{{ v1 is same as (v2) ? 'Yes' : 'No' }}
```

And it returned 3 `Yes` (with PHP 7.2.12)

Same on Twig fiddle: https://twigfiddle.com/bz3lra

**Comment by xabbuh** [comment]
Looks like the YAML parser used by Twig fiddle is responsible for the conversion.

@sparkbuzz Did you only notice this behaviour on Twig fiddle or does it happen in a real application too?

**Comment by fabpot** [comment]
@xabbuh I tried locally on my machine, and it worked well.

**Comment by DarkBee** [comment]
The problem occurs when you are not defining the `value` inside the `twig` file, but passing it from `PHP`.
The fact it is working inside the twig file is because `twig` is turning `0.0` into `0` in the compiler as seen in the source code of the compiled `twig` file

*Controller*

`echo $twig->render('tests/float.html', ['val' => 0.0, ]);`

*Twig file* 

```
{% set val2 = 0.0 %}

{{ val is same as (0.0) ? 'Yes' : 'No' }}
{{ val2 is same as (0.0) ? 'Yes' : 'No' }}
{{ val is same as (val2) ? 'Yes' : 'No' }}

```

*output*

```
No
Yes
No
```

*Compiled template's source*

```
<?php    
protected function doDisplay(array $context, array $blocks = array())
    {
        // line 1
        $context["val2"] = 0;
        // line 2
        $context["val3"] = 0;
        // line 3
        echo "
";
        // line 4
        echo ((((isset($context["val"]) || array_key_exists("val", $context) ? $context["val"] : (function () { throw new Twig_Error_Runtime('Variable "val" does not exist.', 4, $this->source); })()) === 0)) ? ("Yes") : ("No"));
        echo "
";
        // line 5
        echo ((((isset($context["val2"]) || array_key_exists("val2", $context) ? $context["val2"] : (function () { throw new Twig_Error_Runtime('Variable "val2" does not exist.', 5, $this->source); })()) === 0)) ? ("Yes") : ("No"));
        echo "
";
        // line 6
        echo ((((isset($context["val3"]) || array_key_exists("val3", $context) ? $context["val3"] : (function () { throw new Twig_Error_Runtime('Variable "val3" does not exist.', 6, $this->source); })()) === 0)) ? ("Yes") : ("No"));
        echo "

";
        // line 8
        echo ((((isset($context["val"]) || array_key_exists("val", $context) ? $context["val"] : (function () { throw new Twig_Error_Runtime('Variable "val" does not exist.', 8, $this->source); })()) === (isset($context["val2"]) || array_key_exists("val2", $context) ? $context["val2"] : (function () { throw new Twig_Error_Runtime('Variable "val2" does not exist.', 8, $this->source); })()))) ? ("Yes") : ("No"));
        echo "
";
        // line 9
        echo ((((isset($context["val2"]) || array_key_exists("val2", $context) ? $context["val2"] : (function () { throw new Twig_Error_Runtime('Variable "val2" does not exist.', 9, $this->source); })()) === (isset($context["val3"]) || array_key_exists("val3", $context) ? $context["val3"] : (function () { throw new Twig_Error_Runtime('Variable "val3" does not exist.', 9, $this->source); })()))) ? ("Yes") : ("No"));
    }
```


[demo](https://twigfiddle.com/pq5o56)

**Comment by stof** [comment]
Yeah, if we do everything from Twig, things are working because *all* `0.0` are turned into integers, and so no floats are involved at all anymore.

**Comment by fabpot** [comment]
should be fixed by #2795
